api/shared_collection.py

```diff
@@ -0,0 +1,231 @@
+
+    def borrow(self, collection, client, pool, hold=None):
+        api = self.api(collection)
+        self.check_client_authorization(collection, client)
+        if hold and hold.integration_client != client:
+            raise CannotLoan(_("This hold belongs to a different library."))
+        return api.checkout_to_external_library(client, pool, hold=hold)
+
+    def revoke_loan(self, collection, client, loan):
+        api = self.api(collection)
+        self.check_client_authorization(collection, client)
+        if loan.integration_client != client:
+            raise NotCheckedOut(_("This loan belongs to a different library."))
+        return api.checkin_from_external_library(client, loan)
+
+    def fulfill(self, collection, client, loan, mechanism):
+        api = self.api(collection)
+        self.check_client_authorization(collection, client)
+
+        if loan.integration_client != client:
+            raise CannotFulfill(_("This loan belongs to a different library."))
+
+        fulfillment = api.fulfill_for_external_library(client, loan, mechanism)
+        if not fulfillment or not (fulfillment.content_link or fulfillment.content):
+            raise CannotFulfill()
+        
+        if loan.fulfillment is None and not mechanism.delivery_mechanism.is_streaming:
+            __transaction = self._db.begin_nested()
+            loan.fulfillment = mechanism
+            __transaction.commit()
+        return fulfillment
+
+    def place_hold(self, collection, client, pool):
+        api = self.api(collection)
+        self.check_client_authorization(collection, client)
+        return api.place_hold_for_external_library(client, pool)
+
+    def revoke_hold(self, collection, client, hold):
+        api = self.api(collection)
+        self.check_client_authorization(collection, client)
+        if hold and hold.integration_client != client:
+            raise CannotReleaseHold(_("This hold belongs to a different library."))
+        return api.release_hold_from_external_library(client, hold)
+
+
+class BaseSharedCollectionAPI(object):
+    """APIs that permit external circulation managers to access their collections
+    should extend this class."""
+
+    EXTERNAL_LIBRARY_URLS = "external_library_urls"
+
+    SETTINGS = [
+        {
+            "key": EXTERNAL_LIBRARY_URLS,
+            "label": _("URLs for libraries on other circulation managers that use this collection"),
+            "type": "list",
+            "optional": True,
+        },
+        {
+            "key": Collection.EBOOK_LOAN_DURATION_KEY,
+            "label": _("Ebook Loan Duration for libraries on other circulation managers (in Days)"),
+            "default": Collection.STANDARD_DEFAULT_LOAN_PERIOD,
+            "description": _("When a patron uses SimplyE to borrow an ebook from this collection, SimplyE will ask for a loan that lasts this number of days. This must be equal to or less than the maximum loan duration negotiated with the distributor."),
```

Comment:
I would rephrase this as "When a patron of some other library... their circulation manager will ask for a loan..."

For the bit this is based on it's simpler to use "SimplyE" as shorthand for the circulation manager but here we should be more explicit.